### PR TITLE
policy attest

### PR DESCRIPTION
**Pull request by ChrisSchreiber**
- add opa client package
- add grpc evaluate policy method to server
- add grafeas grpc projects clients and update server initialization to work with elasticsearch backend

evaluate policy endpoint is still WIP

### CODE REVIEW

**Review by aalsabag — APPROVED** [review]

File: opa/error.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	errorType ClientErrorType
+	causedBy  error
+}
+
+// ClientErrorType defines error types
+type ClientErrorType string
+
+// OpaClientErrorType constants
+const (
+	OpaClientErrorTypePolicyExits   ClientErrorType = "Policy Exists"
```
[the remark this answers is not in the document]
Comment (ChrisSchreiber) [reply opa/error.go]: 👍 thanks

File: server/server.go [hunk 2]
```diff
@@ -3,28 +3,62 @@ package server
...
 	"go.uber.org/zap"
+	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 	"google.golang.org/protobuf/encoding/protojson"
 )
 
+// NewGrafeasClients construct for GrafeasClients
+func NewGrafeasClients(grafeasEndpoint string) (*GrafeasClients, error) {
+	connection, err := grpc.Dial(grafeasEndpoint, grpc.WithInsecure())
```
Comment (mrparkers) [inline server/server.go:21]: is there a reason that the `DialContext` change I made in #17 was removed here?

File: server/server.go [hunk 3]
```diff
@@ -3,28 +3,62 @@ package server
...
+	if err != nil {
+		return nil, err
+	}
+
+	grafeasClient := grafeas_proto.NewGrafeasV1Beta1Client(connection)
+	projectsClient := grafeas_project_proto.NewProjectsClient(connection)
+
+	return &GrafeasClients{grafeasClient, projectsClient}, err
```
Comment (mrparkers) [inline server/server.go:29]:
```suggestion
	return &GrafeasClients{grafeasClient, projectsClient}, nil
```

File: server/server_test.go [hunk 4]
```diff
@@ -12,205 +12,325 @@ import (
...
 	})
 
-	When("occurrences are created", func() {
-		var (
-			randomOccurrence                      *grafeas_proto.Occurrence
-			grafeasBatchCreateOccurrencesRequest  *grafeas_proto.BatchCreateOccurrencesRequest
-			grafeasBatchCreateOccurrencesResponse *grafeas_proto.BatchCreateOccurrencesResponse
-		)
+	When("server is initialized", func() {
+		It("checks if rode project exists", func() {
```
Comment (mrparkers) [inline server/server_test.go:42]:
```suggestion
		It("should check if the rode project exists", func() {
```

File: server/server_test.go [hunk 5]
```diff
@@ -12,205 +12,325 @@ import (
...
-					randomOccurrence,
-				},
-			}
+			rodeServer, rodeServerError = NewRodeServer(logger.Named("rode server test"), GrafeasClients{grafeasClient, grafeasProjectsClient})
 		})
 
-		It("should forward the batch create occurrence request to grafeas", func() {
-			// ensure Grafeas BatchCreateOccurrences is called with expected request and inject response
-			grafeasClient.EXPECT().BatchCreateOccurrences(gomock.AssignableToTypeOf(context.Background()), gomock.Eq(grafeasBatchCreateOccurrencesRequest)).Return(grafeasBatchCreateOccurrencesResponse, nil)
+		When("rode project does not exists", func() {
```
Comment (mrparkers) [inline server/server_test.go:50]:
```suggestion
		When("the rode project does not exist", func() {
```

File: server/server_test.go [hunk 6]
```diff
@@ -12,205 +12,325 @@ import (
...
+					GetProject(gomock.AssignableToTypeOf(context.Background()), gomock.Eq(getProjectRequest)).
+					Return(nil, status.Error(codes.NotFound, "Not found"))
+			})
 
-			batchCreateOccurrencesRequest := &pb.BatchCreateOccurrencesRequest{
-				Occurrences: []*grafeas_proto.Occurrence{
-					randomOccurrence,
-				},
-			}
+			It("creates rode project", func() {
```
Comment (mrparkers) [inline server/server_test.go:63]:
```suggestion
			It("should create the rode project", func() {
```

File: server/server_test.go [hunk 7]
```diff
@@ -12,205 +12,325 @@ import (
...
+			})
+
+			When("create project returns error from Grafeas", func() {
+				BeforeEach(func() {
+					grafeasProjectsClient.
+						EXPECT().
+						CreateProject(gomock.AssignableToTypeOf(context.Background()), gomock.Eq(createProjectRequest)).Return(nil, status.Error(codes.Internal, "CREATE_PROJECT_ERROR"))
+				})
+
+				It("returns error", func() {
```
Comment (mrparkers) [inline server/server_test.go:78]:
```suggestion
				It("should return an error", func() {
```

File: server/server_test.go [hunk 8]
```diff
@@ -12,205 +12,325 @@ import (
...
-			// check response
-			Expect(response.Occurrences).To(BeEquivalentTo(grafeasBatchCreateOccurrencesResponse.Occurrences))
+			When("create project succeeds", func() {
+				BeforeEach(func() {
+					grafeasProjectsClient.
+						EXPECT().
+						CreateProject(gomock.AssignableToTypeOf(context.Background()), gomock.Eq(createProjectRequest)).Return(&grafeas_project_proto.Project{}, nil)
+				})
+
+				It("returns Rode server", func() {
```
Comment (mrparkers) [inline server/server_test.go:93]:
```suggestion
				It("should return the Rode server", func() {
```

File: server/server_test.go [hunk 9]
```diff
@@ -12,205 +12,325 @@ import (
...
+					rodeServer, rodeServerError = NewRodeServer(logger.Named("rode server test"), GrafeasClients{grafeasClient, grafeasProjectsClient})
+
+					Expect(rodeServer).ToNot(BeNil())
+					Expect(rodeServerError).To(BeNil())
```
Comment (mrparkers) [inline server/server_test.go:97]:
```suggestion
					Expect(rodeServerError).ToNot(HaveOccurred())
```

File: server/server.go [hunk 10]
```diff
@@ -61,9 +95,21 @@ func (r *rodeServer) AttestPolicy(ctx context.Context, request *pb.AttestPolicyR
 	return &pb.AttestPolicyResponse{}, nil
 }
 
-func NewRodeServer(logger *zap.Logger, grafeasClient grafeas.GrafeasV1Beta1Client) pb.RodeServer {
-	return &rodeServer{
-		logger:        logger,
-		grafeasClient: grafeasClient,
+func (r *rodeServer) initialize(ctx context.Context) error {
+	_, err := r.grafeasProjects.GetProject(ctx, &grafeas_project_proto.GetProjectRequest{Name: "projects/rode"})
+	log := r.logger.Named("initialize")
```
Comment (mrparkers) [inline server/server.go:100]: nit: I'd probably make this the first line of the function

File: server/server_test.go [hunk 11]
```diff
@@ -12,205 +12,325 @@ import (
...
-		)
+		When("rode project exists", func() {
+			BeforeEach(func() {
+				grafeasProjectsClient.
+					EXPECT().
+					GetProject(gomock.AssignableToTypeOf(context.Background()), gomock.Eq(getProjectRequest)).
+					Return(&grafeas_project_proto.Project{}, nil)
+			})
+
+			It("does not create project", func() {
```
Comment (mrparkers) [inline server/server_test.go:110]:
```suggestion
			It("should not attempt to create the project", func() {
```

File: server/server_test.go [hunk 12]
```diff
@@ -12,205 +12,325 @@ import (
...
+				grafeasProjectsClient.
+					EXPECT().
+					CreateProject(gomock.Any(), gomock.Any()).MaxTimes(0)
 
+				rodeServer, rodeServerError = NewRodeServer(logger.Named("rode server test"), GrafeasClients{grafeasClient, grafeasProjectsClient})
+			})
+
+			It("returns Rode server", func() {
```
Comment (mrparkers) [inline server/server_test.go:118]:
```suggestion
			It("should return the Rode server", func() {
```

File: server/server_test.go [hunk 13]
```diff
@@ -12,205 +12,325 @@ import (
...
+				rodeServer, rodeServerError = NewRodeServer(logger.Named("rode server test"), GrafeasClients{grafeasClient, grafeasProjectsClient})
+
+				Expect(rodeServer).ToNot(BeNil())
+				Expect(rodeServerError).To(BeNil())
+			})
+		})
+
+		When("get rode project returns error", func() {
```
Comment (mrparkers) [inline server/server_test.go:126]:
```suggestion
		When("fetching the rode project returns an error", func() {
```

File: server/server_test.go [hunk 14]
```diff
@@ -12,205 +12,325 @@ import (
...
+			BeforeEach(func() {
+				grafeasProjectsClient.
+					EXPECT().
+					GetProject(gomock.AssignableToTypeOf(context.Background()), gomock.Eq(getProjectRequest)).
+					Return(nil, status.Error(codes.Internal, "GET_PROJECT_ERROR"))
+			})
+
+			It("returns and error", func() {
```
Comment (mrparkers) [inline server/server_test.go:134]:
```suggestion
			It("should return an error", func() {
```

File: server/server_test.go [hunk 15]
```diff
@@ -12,205 +12,325 @@ import (
...
-			response, err := rodeServer.BatchCreateOccurrences(context.Background(), batchCreateOccurrencesRequest)
-			Expect(err).ToNot(HaveOccurred())
+				rodeServer, rodeServerError = NewRodeServer(logger.Named("rode server test"), GrafeasClients{grafeasClient, grafeasProjectsClient})
+			})
+
+			When("create project returns error from Grafeas", func() {
+				BeforeEach(func() {
+					grafeasProjectsClient.
+						EXPECT().
+						CreateProject(gomock.AssignableToTypeOf(context.Background()), gomock.Eq(createProjectRequest)).Return(nil, status.Error(codes.Internal, "CREATE_PROJECT_ERROR"))
```
Comment (mrparkers) [inline server/server_test.go:75]: can we create or use an existing constant (if one exists) for both `CREATE_PROJECT_ERROR` and `GET_PROJECT_ERROR`?

File: server/server_test.go [hunk 16]
```diff
@@ -12,205 +12,325 @@ import (
...
+				grafeasBatchCreateOccurrencesResponse = &grafeas_proto.BatchCreateOccurrencesResponse{
+					Occurrences: []*grafeas_proto.Occurrence{
+						randomOccurrence,
+					},
+				}
+
+			})
 
-			// expect OPA initPolicy call
+			It("sends occurrences to Grafeas", func() {
```
Comment (mrparkers) [inline server/server_test.go:180]:
```suggestion
			It("should send occurrences to Grafeas", func() {
```

File: server/server_test.go [hunk 17]
```diff
@@ -12,205 +12,325 @@ import (
...
+						}
+						grafeasClient.EXPECT().ListOccurrences(gomock.Any(), gomock.Any()).Return(listOccurrencesResponse, nil)
+
+						// expect evalute OPA policy call
+
+						_, _ = rodeServer.AttestPolicy(context.Background(), attestPolicyRequest)
 					})
-					When("OPA policy evaluation is same as previous attestation occurrence", func() {
+
+					When("resource does not have previous attestation occurrence", func() {
```
Comment (mrparkers) [inline server/server_test.go:251]: there's a decent amount of commented code after this point.  is that needed, or can it be removed?

File: opa/opa_test.go [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+
+					httpmock.RegisterResponder(http.MethodPut, fmt.Sprintf("%s/v1/policies/%s", Opa.Host, opaPolicy),
+						func(req *http.Request) (*http.Response, error) {
+							// todo assert http put body contains policy package
+							return createPolicyResponse, createPolicyError
+						},
+					)
+				})
+
+				JustBeforeEach(func() {
```
Comment (mrparkers) [inline opa/opa_test.go:126]: is this needed?

File: opa/opa_test.go [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+				Expect(initializePolicyError).NotTo(HaveOccurred())
+			})
+		})
+
+		When("policy does not exist", func() {
+			BeforeEach(func() {
+				getPolicyResponse = httpmock.NewStringResponse(404, "{}")
+			})
+
+			It("should fetch policy violations from Elasticsearch", func() {
```
Comment (mrparkers) [inline opa/opa_test.go:105]: is this needed?

File: opa/opa.go [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+func NewClient(logger *zap.Logger, host string) *Client {
+	client := &Client{
+		logger: logger,
+		Host:   host,
+	}
+	return client
+}
+
+// InitializePolicy initializes OPA policy if it does not already exist
+func (opa *Client) InitializePolicy(policy string) ClientError {
```
Comment (mrparkers) [inline opa/opa.go:63]: can we add some whitespace between some of the lines of these functions to make it easier to read?

File: opa/opa.go [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+		log.Error("http request to OPA failed", zap.Error(err))
+		return nil, fmt.Errorf("http request to OPA failed: %s", err)
+	}
+	if httpResponse.StatusCode != http.StatusOK {
+		log.Error("http response status from OPA no OK", zap.Any("status", httpResponse.Status))
+		return nil, fmt.Errorf("http response status not OK: %s", err)
+	}
+
+	response := &EvaluatePolicyResponse{}
+	err = json.NewDecoder(httpResponse.Body).Decode(&response)
```
Comment (mrparkers) [inline opa/opa.go:99]: why `json.NewDecoder().Decode` instead of `json.Unmarshal`?

File: server/server.go [hunk 22]
```diff
@@ -3,28 +3,62 @@ package server
...
+		GrafeasClients: grafeasClients,
+	}
+	if err := rodeServer.initialize(context.Background()); err != nil {
+		return nil, fmt.Errorf("failed to initialize rode server: %s", err)
+	}
+	return rodeServer, nil
+}
+
+// GrafeasClients container for Grafeas protobuf clients
+type GrafeasClients struct {
```
Comment (mrparkers) [inline server/server.go:45]: making a custom type for this seems a bit strange to me - can we just attach the common and project clients to the `rodeServer` struct directly?

**Review by mrparkers — CHANGES_REQUESTED** [review]

File: opa/opa.go (hunk 21 above)
Comment (ChrisSchreiber) [reply opa/opa.go:99]: httpResponse.Body is an io.Reader interface. NewDecoder will read the stream for us. Unmarshal expects a byte array.

File: server/server.go (hunk 22 above)
Comment (ChrisSchreiber) [reply server/server.go:45]: this just reduces the number of arguments to return and pass into the server constructor

Comment (mrparkers) [reply server/server.go:45]: I think I'd prefer if we got rid of it.  I'm not seeing a ton of value using this as an abstraction.  I think it's okay to be explicit with the arguments we pass to the constructor.

File: server/server.go (hunk 2 above)
Comment (ChrisSchreiber) [reply server/server.go]: Nope, missed it when I rebased the changes. I added it back.

File: server/server_test.go (hunk 17 above)
Comment (ChrisSchreiber) [reply server/server_test.go:251]: the comments are placeholders for work to be completed

File: opa/opa_test.go (hunk 19 above)
Comment (ChrisSchreiber) [reply opa/opa_test.go:105]: its a placeholder for functionality what hasn't been added yet

File: server/server.go (hunk 22 above)
Comment (ChrisSchreiber) [reply server/server.go]: 👍

**Review by mrparkers — APPROVED** [review]

**Pull request merged**